core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -161,6 +166,9 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
         if (node.getId().equals(assignedNodeId)) {
             finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
             finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.unassigned() && hasPendingAsyncFetch) {
```

Comment:
for ReplicaShardAllocator, you're looking at the wrong method...
see https://github.com/elastic/elasticsearch/blob/ee4e470f6000fda030ad1f6d9216efec70119227/core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L138